test/add_or_update_test.dart

```diff
@@ -0,0 +1,143 @@
+  late int id;
+  _Friend? bestFriend;
+  final friends = <_Friend>[];
+}
+
+Future<void> main([List<String>? args]) async {
+  await setupTests(args);
+
+  test('Transitive adds', () {
+    final r = getRealm(Configuration.local([Friend.schema]));
+
+    var alice = Friend(1);
+    var bob = Friend(2, bestFriend: alice);
+
+    r.write(() => r.add(bob));
+    expect(bob.isManaged, true);
+    expect(alice.isManaged, true);
+    expect(r.find<Friend>(bob.id), bob);
+    expect(r.find<Friend>(alice.id), alice);
+  });
+
+  test('Transitive updates', () {
+    final r = getRealm(Configuration.local([Friend.schema]));
+
+    var alice = Friend(1);
+    var bob = Friend(2, bestFriend: alice);
+
+    r.write(() => r.add(bob));
+
+    final aliceAgain = Friend(1);
+    final bobAgain = Friend(2, bestFriend: aliceAgain);
+
+    r.write(() => r.add(bobAgain, update: true));
+
+    expect(bob.isManaged, true);
+    expect(alice.isManaged, true);
+    expect(bobAgain.isManaged, true);
+    expect(aliceAgain.isManaged, true);
+
+    // Re-fetch from realm
+    alice = r.find(alice.id)!;
+    bob = r.find(bob.id)!;
+
+    expect(bob, bobAgain);
+    expect(alice, aliceAgain);
+    expect(bob.bestFriend, alice);
+  });
+
+  test('Cycles added correctly', () {
+    final r = getRealm(Configuration.local([Friend.schema]));
+
+    var alice = Friend(1);
+    var bob = Friend(2, bestFriend: alice);
+    alice.bestFriend = bob; // form cycle
+
+    r.write(() => r.add(alice));
+
+    // Re-fetch from realm
+    alice = r.find(alice.id)!;
+    bob = r.find(bob.id)!;
+
+    expect(alice.bestFriend, bob);
+    expect(alice.bestFriend!.bestFriend, alice);
+    expect(alice.bestFriend!.bestFriend!.bestFriend, bob);
+  });
+
+  test('Cycles updated correctly', () {
+    final r = getRealm(Configuration.local([Friend.schema]));
+
+    var alice = Friend(1);
+    var bob = Friend(2, bestFriend: alice);
+    alice.bestFriend = bob; // form cycle
+
+    r.write(() => r.add(alice));
+
+    final aliceAgain = Friend(1);
+    final bobAgain = Friend(2, bestFriend: aliceAgain);
+    aliceAgain.bestFriend = bobAgain;
+
+    alice = r.write(() => r.add(aliceAgain, update: true));
+
+    // Re-fetch from realm
+    alice = r.find(alice.id)!;
+    bob = r.find(bob.id)!;
+
+    expect(alice.bestFriend, bobAgain);
+    expect(alice.bestFriend!.bestFriend, aliceAgain);
+    expect(alice.bestFriend!.bestFriend!.bestFriend, bobAgain);
+  });
+
+  test('Lists updated correctly', () {
+    final r = getRealm(Configuration.local([Friend.schema]));
+
+    var alice = Friend(1);
+    var bob = Friend(2);
+    var carol = Friend(3);
+    alice.friends.addAll([bob, carol]);
+
+    r.write(() => r.add(alice));
+
+    expect(alice.friends, [bob, carol]);
+
+    var dan = Friend(4);
+    final aliceAgain = Friend(1, friends: [dan]);
+
+    r.write(() => r.add(aliceAgain, update: true));
+
+    // Re-fetch from realm
+    alice = r.find(alice.id)!;
+
+    expect(alice.friends, [dan]);
```

Comment:
Is this what the user expects or they expect this element to be added to the existing elements?

Reply:
Updating alice to have a single friend dan, I would be surprised to if either:

alice.friends == [bob, carla, dan], or 
alice.friends == [dan, bob, carla], or
alice.friends == [dan, carla].

I think it is better that the upserted object looks like what was requested after the upsert.